File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.rb
#https://leetcode.com/problems/unique-morse-code-words/discuss/655442/Ruby-solution.
def unique_morse_representations(words)
    a = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
    w_map = ->x{
        x.chars.map{|x| a[x.ord - ?a.ord]}.join
    }
    words.map{|x| w_map[x]}.uniq.size
end

#Third fastest
# @param {String[]} words
# @return {Integer}
def unique_morse_representations(words)
    map = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
    
    ret = []
    words.each_with_index do |word, idx|
      ret[idx] = ''
      bytes = word.bytes.map { |x| x - 97 }
      bytes.each { |i| ret[idx] << map[i] }
    end
    
    ret.uniq.size
  end

#Second fastest
# @param {String[]} words
# @return {Integer}
def unique_morse_representations(words)
    morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
    alpha = "abcdefghijklmnopqrstuvwxyz"
    hash = {}
    unique = []
    alpha.each_char.with_index {|char, idx| hash[char] = morse[idx]}
    words.each do |word|
        new = ""
        word.each_char do |char|
            new += hash[char]
        end
        unique<<new
    end
    return unique.uniq.count
end

#Fastest
# @param {String[]} words
# @return {Integer}
def unique_morse_representations(words)
    morses = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
    ret = {}
    words.each do |word|
      mos = ""
      word.each_byte do |chr|
        mos = mos + morses[chr-97]
      end
      ret[mos] = true
    end
    ret.keys.size
  end

File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.py
#Fourth fastest
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        MORSE = [".-","-...","-.-.","-..",".","..-.","--.",
                 "....","..",".---","-.-",".-..","--","-.",
                 "---",".--.","--.-",".-.","...","-","..-",
                 "...-",".--","-..-","-.--","--.."]

        temp = []
        for word in words:
            t = ""
            for c in word: 
                t += MORSE[ord(c)-ord('a')]
            temp.append(t)
        return len(set(temp))

#Third fastest
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
      dic = {}
      morse = [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
      for idx, m in enumerate(morse):
        dic[chr(idx+97)] = m
      
      result = []
      for mess in words:
        tr = ""
        for ch in mess:
          tr += dic[ch]
        result += [tr]
      
      return len(set(result))

#Second fastest
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        m = {}
        
        for x, y in zip(
            [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."],
            'abcdefghijklmnopqrstuvwxyz'
        ):
            m[y] = x
        
        result = set()
        
        for word in words:
            result.add(
                ''.join([m[c] for c in word])
            )
        
        return len(result)


#Fastest
# Platform: leetcode.com
# No. 804. Unique Morse Code Words
# Link: https://leetcode.com/problems/unique-morse-code-words/
# Difficulty: Easy
# Dev: Chumicat
# Date: 2020/11/15
# Submission: https://leetcode.com/submissions/detail/420221386/
# (Time, Space) Complexity : O(n), O(n)
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:
        table = {'i': '..', 's': '...', 'd': '-..', 'h': '....', 'l': '.-..', 'b': '-...', 
                 'a': '.-', 'u': '..-', 'k': '-.-', 'v': '...-', 'z': '--..', 'x': '-..-', 
                 'n': '-.', 'r': '.-.', 'g': '--.', 'f': '..-.', 'p': '.--.', 'c': '-.-.', 
                 'm': '--', 'w': '.--', 'o': '---', 'q': '--.-', 'j': '.---', 'y': '-.--', 
                 'e': '.', 't': '-',}
        return len({"".join(table[c] for c in word) for word in words})
